metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py

```diff
@@ -0,0 +1,814 @@
+
+    def lineage_urn_to_lowercase(self, value):
+        return Mapper.urn_to_lowercase(
+            value, self.__config.convert_lineage_urns_to_lowercase
+        )
+
+    def assets_urn_to_lowercase(self, value):
+        return Mapper.urn_to_lowercase(value, self.__config.convert_urns_to_lowercase)
+
+    def new_mcp(
+        self,
+        entity_type,
+        entity_urn,
+        aspect_name,
+        aspect,
+        change_type=ChangeTypeClass.UPSERT,
+    ):
+        """
+        Create MCP
+        """
+        return MetadataChangeProposalWrapper(
+            entityType=entity_type,
+            changeType=change_type,
+            entityUrn=entity_urn,
+            aspectName=aspect_name,
+            aspect=aspect,
+        )
+
+    def __to_work_unit(
+        self, mcp: MetadataChangeProposalWrapper
+    ) -> EquableMetadataWorkUnit:
+        return Mapper.EquableMetadataWorkUnit(
+            id="{PLATFORM}-{ENTITY_URN}-{ASPECT_NAME}".format(
+                PLATFORM=self.__config.platform_name,
+                ENTITY_URN=mcp.entityUrn,
+                ASPECT_NAME=mcp.aspectName,
+            ),
+            mcp=mcp,
+        )
+
+    def __to_datahub_dataset(
+        self, dataset: Optional[PowerBiAPI.PowerBIDataset]
+    ) -> List[MetadataChangeProposalWrapper]:
+        """
+        Map PowerBi dataset to datahub dataset. Here we are mapping each table of PowerBi Dataset to Datahub dataset.
+        In PowerBi Tile would be having single dataset, However corresponding Datahub's chart might have many input sources.
+        """
+
+        dataset_mcps: List[MetadataChangeProposalWrapper] = []
+        if dataset is None:
+            return dataset_mcps
+
+        LOGGER.info(
+            f"Converting dataset={dataset.name}(id={dataset.id}) to datahub dataset"
+        )
+
+        for table in dataset.tables:
+            # Create a URN for dataset
+            ds_urn = builder.make_dataset_urn(
+                platform=self.__config.platform_name,
+                name=self.assets_urn_to_lowercase(table.full_name),
+                env=self.__config.env,
+            )
+
+            LOGGER.info(f"{Constant.Dataset_URN}={ds_urn}")
+            # Create datasetProperties mcp
+            ds_properties = DatasetPropertiesClass(
+                name=table.name, description=table.name
+            )
+
+            info_mcp = self.new_mcp(
+                entity_type=Constant.DATASET,
+                entity_urn=ds_urn,
+                aspect_name=Constant.DATASET_PROPERTIES,
+                aspect=ds_properties,
+            )
+
+            # Remove status mcp
+            status_mcp = self.new_mcp(
+                entity_type=Constant.DATASET,
+                entity_urn=ds_urn,
+                aspect_name=Constant.STATUS,
+                aspect=StatusClass(removed=False),
+            )
+            dataset_mcps.extend([info_mcp, status_mcp])
+
+            if self.__config.extract_lineage is True:
```

Comment:
Starting at this line this methods get a bit too long. Can we split into more reasonably sized functions?

Reply:
Moved it to a separate function.